src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -165,6 +165,7 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase? conversions, N
         internal static InferredLambdaReturnType InferReturnType(ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)> returnTypes,
             UnboundLambda node, Binder binder, TypeSymbol? delegateType, bool isAsync, ConversionsBase conversions)
         {
+            Debug.Assert(!node.HasExplicitReturnType(out _, out _));
```

Comment:
Consider moving assertion into `InferReturnTypeImpl` so that it also covers the `InferReturnType` overload a few lines above (line 159)

Reply:
Ping in case you missed this suggestion

Reply:
Thanks. There's more work required to move this assert. I'll investigate.